prometheus-metrics-model/src/main/java/io/prometheus/metrics/model/snapshots/MetricMetadata.java

```diff
@@ -105,5 +105,11 @@ private void validate() {
             throw new IllegalArgumentException("'" + name + "': Illegal metric name. " + error
                     + " Call " + PrometheusNaming.class.getSimpleName() + ".sanitizeMetricName(name) to avoid this error.");
         }
+        if (hasUnit()) {
+            if (!name.endsWith("_" + unit) && !name.endsWith("." + unit)) {
+                throw new IllegalArgumentException("'" + name + "': Illegal metric name. The name must end with _" + unit + "."
```

Comment:
What do you think about adding some extra information in the error message?
```suggestion
                throw new IllegalArgumentException("'" + name + "': Illegal metric name. If the unit is non-null, the name must end with it (_" + unit + ")."
```

Reply:
Also, there is a twist: this is only a problem for the OpenMetrics format, for the Prometheus Text format it is completely ok doing this (since there is no `# UNIT` anyways). I think though for the sake of consistency this should be validated in both cases to avoid client-side logic that produces different names if the format is different.

Reply:
Thanks, I adapted the Exception message.

I agree, while it's theoretically possible to expose the metric name without the unit suffix in text format this is not something we should support.